### PR TITLE
Repaired some of ListenableFuture.kt's cancellation corner cases

### PR DESCRIPTION
**Pull request by yorickhenning**
This fixes:
  - Cancellation without an untrapped CancellationException propagating
    through a Callback; isCancelled() is the correct way to check for
    cancellation
  - Bidirectional propagation of cancellation through
    `asListenableFuture()`
  - The cause getting lost in the `asListenableFuture()` future  when
    cancelling its `Deferred` parent with a cause

This also:
  - Extensively documents the package and the contracts created by the
    promise-creating extension methods and `future()`
  - Uses `getUninterruptibly()` for speed
  - Uses `AbstractFuture` to make as certain as possible that
    `Future.cancel()` will return `true` at most once
  - Should clear up rare spooky race conditions around
    cancellation/interruption in hybrid Coroutines/Guava Futures
    codebases

There are probably a few more interesting corner cases hiding in here,
but this should be a good start improving the correctness of `.guava`'s
adapters.

### CODE REVIEW

**Comment by yorickhenning** [comment]
@charlesmunger
@cpovirk

**Comment by yorickhenning** [comment]
Should resolve #1346

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 1]
```diff
@@ -5,31 +5,51 @@
...
+ * See [newCoroutineContext][CoroutineScope.newCoroutineContext] for a description of debugging
+ * facilities.
  *
- * @param context additional to [CoroutineScope.coroutineContext] context of the coroutine.
+ * Note that the error semantics of [future] are _subtly different_ than [asListenableFuture].
+ * By calling [future], any error that occurs after successfully cancelling the returned
+ * [ListenableFuture] will be passed to the [CoroutineExceptionHandler] from the context. The
+ * contract of [Future] does not permit it to return an error after it is successfully cancelled.
+ *
+ * By calling [asListenableFuture] on a [Deferred], any error that occurs after successfully
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: This paragraph is not necessary for the public API, it's better to move to implementation

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 2]
```diff
@@ -38,81 +58,63 @@ public fun <T> CoroutineScope.future(
...
+    if (isDone && !isCancelled) {
         return try {
-            @Suppress("UNCHECKED_CAST")
-            CompletableDeferred(get() as T)
-        } catch (e: Throwable) {
-            // unwrap original cause from ExecutionException
-            val original = (e as? ExecutionException)?.cause ?: e
-            CompletableDeferred<T>().also { it.completeExceptionally(original) }
+            // getUninterruptibly() avoids one volatile read over Futures.getDone().
+            CompletableDeferred(Uninterruptibles.getUninterruptibly(this))
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Guava has a bad reputation with `@Beta` API compatibility. It's better to use stable API, especially if it only costs one volatile read

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt [hunk 3]
```diff
@@ -205,6 +336,59 @@ class ListenableFutureTest : TestBase() {
...
+        val asDeferredAsFuture = asDeferred.asListenableFuture()
+
+        assertTrue(asDeferredAsFuture.isCancelled)
+        assertFailsWith<CancellationException> {
+            val value: Int = asDeferredAsFuture.await()
+        }
+    }
+
+    @Test
+    fun testCancelledDeferredAsListenableFutureAsDeferredJoinDoesNotThrow() = runTest {
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt]: `join` throws if and only if a caller of the `join` was cancelled

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt [hunk 4]
```diff
@@ -311,9 +495,9 @@ class ListenableFutureTest : TestBase() {
 
     @Test
     fun testUnhandledExceptionOnExternalCancellation() = runTest(
-        unhandled = listOf(
-            { it -> it is TestException } // exception is unhandled because there is no parent
-        )
+      unhandled = listOf(
+        { it -> it is TestException } // exception is unhandled because there is no parent
+      )
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt]: this is not the library code-style, please revert this change

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 2 above)
Comment (charlesmunger) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: This method is not `@Beta`:

https://github.com/google/guava/blob/master/guava/src/com/google/common/util/concurrent/Uninterruptibles.java#L232

Comment (qwwdfsad) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Good to know, apparently we use a slightly outdated version of Guava

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 5]
```diff
@@ -129,36 +131,289 @@ public fun <T> ListenableFuture<T>.asDeferred(): Deferred<T> {
...
+private class ListenableFutureCoroutine<T>(
+    context: CoroutineContext,
+    private val future: SettableFuture<T>
+) : AbstractCoroutine<T>(context)  {
+
+    override fun onCompleted(value: T) {
+        future.set(value)
+    }
+
+    // TODO: This doesn't actually cancel the Future. There doesn't seem to be bidi cancellation?
```
[the remark this answers is not in the document]
Comment (charlesmunger) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: A Future has four states:

1. Pending (isDone == false, isCancelled == false, get will block)
2. Cancelled (isDone == true, isCancelled == true, get will throw CancellationException)
3. Failed (isDone == true, isCancelled == false, get will throw ExecutionException)
4. Succeeded (isDone == true, isCancelled == false, get will return a value)

Thus there is a distinction between "the future is cancelled, and throws CancellationException from get()" and "The future is failed, and thus will throw an ExecutionException with a cause of CancellationException from get()"

A Future should only throw CancellationException from get() if isCancelled() returns true, and vice versa. 

This distinction doesn't seem to exist with Deferred, and so there's a bit of a mismatch. This behavior is what we came up with to best bridge the gap; if the Deferred itself is cancelled (by the Deferred's definition, which is failing by CancellationException), then the future reports itself is cancelled. 

This java code:
```java
Future<?> future = executorService.submit(() -> {throw new CancellationException();});
future.get();
```

Will throw an ExecutionException with a cause of CancellationException. That behavior matches your latter case.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 6]
```diff
@@ -38,81 +58,63 @@ public fun <T> CoroutineScope.future(
...
+ *
+ * When either promise successfully completes, it will attempt to synchronously complete its
+ * counterpart with the same value. This will succeed barring a race with cancellation.
+ *
+ * When either promise completes with an Exception, it will attempt to synchronously complete its
+ * counterpart with the same Exception. This will succeed barring a race with cancellation.
+ *
+ * Cancellation is propagated bidirectionally.
+ *
+ * When `this` [ListenableFuture is successfully cancelled - meaning [ListenableFuture.cancel]
```
[the remark this answers is not in the document]
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Done.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 7]
```diff
@@ -38,81 +58,63 @@ public fun <T> CoroutineScope.future(
...
-
 /**
- * Converts this listenable future to an instance of [Deferred].
- * It is cancelled when the resulting deferred is cancelled.
+ * Returns a [Deferred] that is completed or failed by `this` [ListenableFuture].
+ *
+ * Completion is non-atomic between the two promises.
+ *
+ * When either promise successfully completes, it will attempt to synchronously complete its
+ * counterpart with the same value. This will succeed barring a race with cancellation.
```
[the remark this answers is not in the document]
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: By "completes" I mean "reaches Cancelled or Completed state" for a `Deferred`, or "finishes" for a `Future`. I'm referring to the promise ending, not its being completed in the sense of calling `CompletableDeferred.complete()`.

Is there another way I can phrase this to make that intention clearer?

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt (hunk 4 above)
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt]: Autoformatter strikes, I think. Unintentional diff. Reverted.

Is there a documented style? I'm working from IntelliJ's autoformatter.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 1 above)
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Moved to `CancellationToCoroutine` class doc and crosslinked.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Done.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 8]
```diff
@@ -38,81 +58,63 @@ public fun <T> CoroutineScope.future(
 ): ListenableFuture<T> {
     require(!start.isLazy) { "$start start is not supported" }
     val newContext = newCoroutineContext(context)
+    // TODO: It'd be nice not to leak this SettableFuture reference, which is easily blind-cast.
     val future = SettableFuture.create<T>()
     val coroutine = ListenableFutureCoroutine(newContext, future)
-    Futures.addCallback(future, coroutine, MoreExecutors.directExecutor())
+    future.addListener(
+      CancellationToCoroutine(future, coroutine),
```
[the remark this answers is not in the document]
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: I noticed that optimization but figured it might be a bit obtuse to open with it. :)

Done.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 9]
```diff
@@ -38,81 +58,63 @@ public fun <T> CoroutineScope.future(
...
+            // ExecutionException is the only kind of exception that can be thrown from a gotten
+            // Future. Anything else showing up here indicates a very fundamental bug in a
+            // Future implementation.
+            CompletableDeferred<T>().also { it.completeExceptionally(e.nonNullCause()) }
         }
     }
+
     val deferred = CompletableDeferred<T>()
     Futures.addCallback(this, object : FutureCallback<T> {
         override fun onSuccess(result: T?) {
```
[the remark this answers is not in the document]
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: True. Good catch.

In my first read, I thought the unchecked cast was equivalent to the !! assert. Re-reading, the previous behaviour was to catch and `completeExceptionally()` with a KNPE. I've restored that behaviour with a version that won't try/catch for the null case. I've added a test and documentation to make sure the behaviour doesn't change unnoticed again.

The fundamental problem is that `ListenableFuture<T>` can complete with a null value, and `CompletableDeferred<T>` (and Deferred<T>) cannot. Offhand, I don't think there's a better solution than this. Using the type `LF<T?>` in the method seems like a high cost.

I started addressing cancellation, not nullability, but here we are with bonus test cases. :)

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt (hunk 3 above)
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt]: Am I understanding right that you mean the test case's name is redundant? I think I see tha. I've picked a new name to make it clearer what I'm trying to test for here.

To paraphrase, my intention is to test that the correct `Deferred` cancellation completion behaviour tunnels through a `Deferred` -> `Future` -> `Deferred` transformation. That's a good guarantee that the two transforms interoperate properly.

Let me know if I've misunderstood your perspective here.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 1 above)
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Move to `ListenableFutureCoroutine` after merging `CancellationToCoroutine`.

**Comment by yorickhenning** [comment]
Any further feedback? I'd like to get this finished up.

**Comment by qwwdfsad** [comment]
Sorry, I'm busy with 1.3 release and internal activities, will return to your PR on Friday

**Comment by yorickhenning** [comment]
> Sorry, I'm busy with 1.3 release and internal activities, will return to your PR on Friday

No worries! Just wanted to make sure it's not forgotten.

File: integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt (hunk 4 above)
Comment (qwwdfsad) [reply integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt]: There isn't, but we are also using IntelliJ's autoformatter ¯\_(ツ)_/¯

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 10]
```diff
@@ -77,88 +292,143 @@ private class ListenableFutureCoroutine<T>(
...
+ *     [ListenableFuture]'s documented and implicit contracts is surprisingly difficult to achieve.
+ *     The best way to be correct, especially given the fun corner cases from
+ *     [AsyncFuture.setAsync], is to just use an [AsyncFuture].
+ *   - To maintain sanity, this class implements [ListenableFuture] and uses an inner [AsyncFuture]
+ *     around its input [deferred] as a state engine to establish happens-after-completion. This
+ *     could probably be compressed into one subclass of [AsyncFuture] to save an allocation, at the
+ *     cost of the implementation's readability.
  */
-public fun <T> Deferred<T>.asListenableFuture(): ListenableFuture<T> = DeferredListenableFuture<T>(this)
+private class OuterFuture<T>(val deferred: Deferred<T>): ListenableFuture<T> {
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: `private val deferred` will save two methods in the bytecode :)

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 11]
```diff
@@ -77,88 +292,143 @@ private class ListenableFutureCoroutine<T>(
...
+ *
+ * - Correct implementation of [Future]'s happens-after semantics documented for [get], [isDone]
+ *   and [isCancelled] methods
+ * - Cancellation propagation both to and from [Deferred]
+ * - Correct cancellation and completion semantics even when this [ListenableFuture] is combined
+ *   with different concrete implementations of [ListenableFuture]
+ *   - Fully correct cancellation and listener happens-after obeying [Future] and
+ *     [ListenableFuture]'s documented and implicit contracts is surprisingly difficult to achieve.
+ *     The best way to be correct, especially given the fun corner cases from
+ *     [AsyncFuture.setAsync], is to just use an [AsyncFuture].
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: There is no such class as `AsyncFuture`

**Comment by qwwdfsad** [comment]
Almost there :)
In the same time, I will start to do the same exercise for `CompletableFuture`

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 10 above)
Comment (JakeWharton) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Is there a YouTrack for fixing this to happen automatically? Even if it's in the new IR backend only, you shouldn't be required to know things like this.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 11 above)
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: AbstractFuture. Not sure how I got that one wrong.

Fixed.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 10 above)
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Done.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 12]
```diff
@@ -38,28 +45,235 @@ public fun <T> CoroutineScope.future(
...
+ * the task is cancelled via the [Deferred] reference to it. This is unavoidable, so make sure
+ * no idempotent work is performed by any users of the [ListenableFuture] reference to the task
+ * if they're the thread that succeeded in cancelling the task - it may not get done.
+ *
+ * This is inherently a race. See [Future.cancel] for a description of `Future` cancellation
+ * semantics. See [Job] for a description of coroutine cancellation semantics.
+ */
+public fun <T> ListenableFuture<T>.asDeferred(): Deferred<T> {
+    // Fast path if already completed.
+    // TODO: There's a faster fast path - InternalFutureFailureAccess - in Guava >= 27.0
```
[the remark this answers is not in the document]
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Done. Will send another PR with the latest requested changes.

Let's see if I still know how to hold git right...

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 10 above)
Comment (elizarov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: @JakeWharton See https://youtrack.jetbrains.com/issue/KT-16454

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (yorickhenning) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Done. Moved the details inside the method as a block comment.

**Comment by yorickhenning** [comment]
Rebase on develop w/ some improvements: https://github.com/Kotlin/kotlinx.coroutines/pull/1441
